apps/tlon-web/src/dms/BroadcastInviteDialog.tsx

```diff
@@ -0,0 +1,160 @@
+import { useCallback, useState } from 'react';
+import { useNavigate } from 'react-router';
+
+import Dialog from '../components/Dialog';
+import ShipSelector, { ShipOption } from '../components/ShipSelector';
+import { refetchCohort, useCohort, useCohorts } from '@/state/broadcasts';
+import api from '@/api';
+import { stringToTa } from '@/logic/utils';
+
+interface BroadcastInviteDialogProps {
+  inviteIsOpen: boolean;
+  setInviteIsOpen: (open: boolean) => void;
+  whom: string;
+  mode: 'add' | 'del';
+  create: boolean;
+}
+
+export default function BroadcastInviteDialog({
+  inviteIsOpen,
+  setInviteIsOpen,
+  whom,
+  mode = 'add',
+  create = false,
+}: BroadcastInviteDialogProps) {
+  const navigate = useNavigate();
+  const [ships, setShips] = useState<ShipOption[]>([]);
+  const targets = useCohort(whom).targets;
+  const invalidShips = ships.filter((ship) => {
+    if (!targets) {
+      return false;
+    }
+    if (mode === 'add') {
+      return targets.includes(ship.value);
+    } else {
+      return !targets.includes(ship.value);
+    }
+  });
+  const showError = invalidShips.length > 0;
+  const [nameValue, setNameValue] = useState('');
+  const { refetch: refetchCohorts } = useCohorts();
+
+  const onEnter = useCallback(async () => {
+    navigate(`/dm/broadcasts/${whom}`);
+  }, [navigate, whom]);
+
+  const submitHandler = useCallback(async () => {
+    if (create && nameValue) {
+      const ta = stringToTa(nameValue);
+      const json = {
+        'add-cohort': {
+          cohort: ta,
+          targets: ships.map((so) => {
+            return so.value;
+          })
+        }
+      };
+      const after = () => {
+        refetchCohorts();
+        navigate(`/dm/broadcasts/${ta}`);
+        setInviteIsOpen(false);
+        setNameValue('');
+        setShips([]);
+      };
+      api.poke({
+        mark: 'broadcaster-action', app: 'broadcaster', json,
+        onSuccess: ()=>after(), onError: ()=>after()
+      });
+    } else
+    if (whom && !showError) {
+      let json;
+      if (mode === 'add') {
+        json = {
+          'add-cohort': {
+            cohort: whom,
+              targets: ships.map((so) => {
+                return so.value;
+              })
+          }
+        };
+      } else {
+        json = {
+          'del-cohort': {
+            cohort: whom,
+              targets: ships.map((so) => {
+                return so.value;
+              })
+          }
+        };
+      }
+      //TODO  refetch just this specific cohort
+      api.poke({
+        mark: 'broadcaster-action', app: 'broadcaster', json,
+        onSuccess: refetchCohorts, onError: refetchCohorts
+      });
```

Comment:
Normally all of this should live outside of the component, in this case it would probably go in state/broadcasts.